iothub/device/src/Transport/Amqp/AmqpUnit.cs

```diff
@@ -496,40 +553,21 @@ public async Task SendTwinMessageAsync(AmqpTwinMessageType amqpTwinMessageType,
         public void OnConnectionDisconnected()
         {
             if (Logging.IsEnabled) Logging.Enter(this, $"{nameof(OnConnectionDisconnected)}");
-            if (SetNotUsable() == 0)
-            {
-                _amqpAuthenticationRefresher?.StopLoop();
-                OnUnitDisconnected?.Invoke(false, EventArgs.Empty);
-            }
-
+            _amqpAuthenticationRefresher?.StopLoop();
+            _onUnitDisconnected();
+            
             if (Logging.IsEnabled) Logging.Exit(this, $"{nameof(OnConnectionDisconnected)}");
         }
 
         private void OnSessionDisconnected(object o, EventArgs args)
         {
             if (Logging.IsEnabled) Logging.Enter(this, o, $"{nameof(OnSessionDisconnected)}");
-
-            if (SetNotUsable() == 0)
+            if (ReferenceEquals(o, _amqpIoTSession))
```

Comment:
This is unclear: why would it not be the same session?
Can you have two sessions at the same time linked to this callback? If yes, the referenceEquals check is wrong anyway as there is no interlocking on the objects which means that different threads may get different values.

Reply:
The equals comparison will ensure that the event is only for current session and hence handle the race condition. By design.